Upgrade to Astro v3 guide (#4166)
withastro/docs: src/content/docs/en/guides/upgrade-to/v3.mdx

 What should I do?
 If you previously avoided the assets feature due these constraints, you can now use them without issues. You can also use the no-op image service to explicitly opt-in to this behaviour:

 ```
 // astro.config.mjs
 export default {
   image: {
     service: {
       entrypoint: 'astro/assets/services/noop'
     }
   }
 }
 ```

In Astro v2.x, you would get an error log in development. When building a project, the build would fail.

Astro V3 still displays the error log, but now you can build the project.

When using an adapter that supports neither Squoosh or Sharp, Astro will now automatically use an image service that does not support processing, but still provides the other benefits of `astro:assets` such as enforcing `alt`, no CLS etc to users. Should be noted that it does no transformations or processing, but you still get all the other benefits (no CLS, alt enforced, etc). All of those benefits are part of the image service, if you make your own, you lose those benefits unless you make it yourself

> ## Known Issues
> 
> There are currently no known issues.

Reviewer: ``````suggestion
### Removed: `<Markdown />` component

In Astro v2.x, Astro deprecated the `<Markdown />` component and moved it to an external package.

Astro v3.0 comletely removes the package `@astrojs/markdown-component`, so Astro's `<Markdown />` component will no longer work in your project.

#### What should I do?

Uninstall the `@astrojs/markdown-component` package. To continue using a similar `<Markdown />` component in your code, consider using [community integrations](https://astro.build/integrations/) that provide a similar component like https://github.com/natemoo-re/astro-remote, and be sure to update your component imports and attributes as necessary according to the integration's own documentation. Otherwise, delete all references to importing Astro's `<Markdown />` component and the component itself in your `.astro` files. You will need to rewrite your content as HTML directly or [import Markdown](/en/guides/markdown-content/#importing-markdown) from an `.md` file.

### Removed: camelCase transformations

In Astro 2.x, Astro automatically transformed camelCase CSS variable names passed to the `style` attribute of an HTML element.

Astro 3.0 removes backwards-compatible kebab-case transform for these camelCase CSS variable names. 

#### What should I do?

If you were relying on Astro to transform kebab-case in your styles, update your existing styles to use the camelCase version to prevent missing styles. For example:

```astro
---
const myValue = "red"
---
<!-- input -->
<div style={{ "--myValue": myValue }}></div>
<!-- output (before) -->
<div style="--my-value:var(--myValue);--myValue:red"></div>
<!-- output (after) -->
<div style="--myValue:red"></div>
```

```diff
<style>
  div {
-   color: var(--my-value);
+   color: var(--myValue);
  }
</style>
```

## Known Issues

There are currently no known issues.

``````

In Astro v2.x, Astro offered an official image integration that included Astro `<Image />` and `<Picture />` components.

Astro v3.0 removes this integration from the codebase entirely. Astro's new solution for images is a built-in image services API: `astro:assets`.

#### What should I do?

Remove the `@astrojs/image` integration from your project. You will need to not only uninstall the integration, but also update or remove any import statements and existing `<Image />` and `<Picture />` components. You might also need to configure a preferred default image processing service.

You will find [complete, step-by-step instructions for removing the old image integration](/en/guides/images/#remove-astrojsimage) in our Images guide.

Migrating to `astro:assets` will also bring some new image options and features that you may now wish to use. Please see the full [v3.0 Image Upgrade Advice](/en/guides/images/#upgrade-to-v30-from-v2x) for full details!

```js del={3,7}
// astro.config.mjs
import { defineConfig } from 'astro/config';
import image from '@astrojs/image';

export default defineConfig({
  integrations: [
    image(),
  ]
})
```

### Removed: `<Markdown />` component

> In Astro v2.x, Astro deprecated the `<Markdown />` component and moved it to an external package.

Reviewer 2: I think I made a mistake in my entry. The Markdown component is deprecated since Astro v1
```suggestion
In Astro v1.x, Astro deprecated the `<Markdown />` component and moved it to an external package.
```

### Changed default: port `3000`

In Astro v2.x, Astro ran on port `3000` by default.

Astro v3.0 changes the [default port](/en/reference/cli-reference/#--port-number) to `4321`. 🚀

#### What should I do?

Update any existing references to `localhost:3000`, for example in tests or in your `README`, to reflect the new port `localhost:4321`.

### Changed default: import.meta.env.BASE_URL `trailingSlash`

In Astro v2.x, `import.meta.env.BASE_URL` appended your [`base`](/en/reference/configuration-reference/#base) setting with a [trailingSlash](/en/reference/configuration-reference/#trailingslash) by default. `trailingSlash: "ignore"`also appended a trailing slash.

Astro v3.0 no longer appends `import.meta.env.BASE_URL` with a trailing slash by default, nor when `trailingSlash: "ignore"` is set. (The existing behavior of `base` in combination with `trailingSlash: "always"` or `trailingSlash: "never"` is unchanged.)

#### What should I do?

If your `base` already has a trailing slash, no change is needed.

> If your `base` does not have a trailing slash, add one if you wish to preserve the previous default (or `trailingSlash: "ignore"`) behaviour:

Reviewer 3: 🚓 British English Police 👮 
```suggestion
If your `base` does not have a trailing slash, add one if you wish to preserve the previous default (or `trailingSlash: "ignore"`) behavior:
```

Astro v3.0 now compresses emitted HTML by default.

#### What should I do?

You can now remove `compressHTML: true` from your configuration as this is the new default behavior.

```js title="astro.config.mjs" del={4}
import { defineConfig } from "astro/config";

export default defineConfig({
	compressHTML: true
})
```

You must now set `compressHTML: false` to opt out of HTML compression.

### Changed default: `scopedStyleStrategy`

In Astro v2.x, the default value of [`scopedStyleStrategy`](/en/reference/configuration-reference/#scopedstylestrategy) was `"where"`.

Astro v3.0 introduces a new, default value: `"attribute"`. By default, styles are now applied using `data-*` attributes.

#### What should I do?

To retain your project's current [style scoping](/en/guides/styling/#scoped-styles), update the configuration file to the previous default value: 

```js title="astro.config.mjs" ins={4}
import { defineConfig } from "astro/config";

export default defineConfig({
	scopedStyleStrategy: "where"
})
```

### Changed default: `inlineStyleSheets`

> In Astro v2.x, all project stylesheets were sent as link tags by default. You could opt in to `"always"` inlining them into `<style>` tags, or to `"auto"`matically inlining only stylesheets below a certain size by setting the [`build.inlineStylesheets`](/en/reference/configuration-reference/#buildinlinestylesheets) configuration. The default setting was `"never"`.

Reviewer 3: This is `"auto"`matically asking for translators to rack their brains 🤣

Super fun and smart paragraph crafting, but I feel like I have to destroy your joy!
```suggestion
In Astro v2.x, all project stylesheets were sent as link tags by default. You could opt in to inlining them into `<style>` tags everytime with `"always"`, or to inlining only stylesheets below a certain size with `"auto"` by setting the [`build.inlineStylesheets`](/en/reference/configuration-reference/#buildinlinestylesheets) configuration. The default setting was `"never"`.
```

Reviewer: Worthy of a screenshot, @Yan-Thomas! I have taught you well.

Astro v3.0 changes the default value of `inlineStylesheets` to `"auto"`. Stylesheets smaller than `ViteConfig.build.assetsInlineLimit` (default: 4kb) are inlined by default. Otherwise, project styles are sent in external stylesheets.

#### What should I do?
If you want to keep your project's current behavior, set `build.inlineStylesheets` to the previous default, `"never"`:

```js title="astro.config.mjs" ins={4-6}
import { defineConfig } from "astro/config";

export default defineConfig({
	 build: {
    inlineStylesheets: "never"
  }
})
```

### Changed default: image service

In Astro v2.x, Squoosh was the [default image processing service](/en/guides/images/#default-image-service).

> Astro v3.0 now includes Sharp as the default image processing service, and instead provides a configuration option to use Squoosh.

Reviewer 3: Comma IX - The Rise of Skywalker
```suggestion
Astro v3.0 now includes Sharp as the default image processing service and instead provides a configuration option to use Squoosh.
```

#### What should I do?

> If you had previously installed Sharp, uninstall the dependency using your package manager, or by removing manually from your `package.json`. It is now included in Astro.

Reviewer 4: I don't believe removal of `Sharp` should be recommended. See https://github.com/withastro/astro/issues/7966 and related issue.

Reviewer: Thanks for pointing this out, @fflaten ! Would you like to suggest a change in wording here? Do you think it's safe to say "you MAY be able to remove Sharp", or would you instead suggest NO mention at all?

Also asking @Princesseuh for her advice here. Is there any harm in having this as a separate, manually installed dependency if Astro DOES successfully do it for you?

Happy to let you two figure out the best advice here, then I'll just commit it!

Reviewer 4: I'm very new to web dev, but AFAIK there's no clear downside with keeping it as a direct dependency. If true, then I'd leave it out.

Similarly I'd keep it out of install docs and let the error message help any affected users.

Reviewer 5: No downside keeping it, in fact `pnpm` users even need it directly installed in their projects, so I'd say might as well.

The only potential downside is their sharp install being duped, but we try to maintain a good enough semver range that this shouldn't happen.

- `del` to `DELETE`

```js title="endpoint.ts" del={1} ins={2}
export function get() {
export function GET() {
    return new Response(JSON.stringify({ "title": "Bob's blog" }));
}
```

### Changed: Multiple JSX framework configuration

In Astro v2.x, you could use [multiple JSX framework integrations](/en/guides/integrations-guide/#official-integrations) (React, Solid, Preact) in the same project without needing to identify which files belonged to which framework.

Astro v3.0 now requires you to specify which framework to use for your files with new `include` and `exclude` integration config options when you have multiple JSX framework integrations installed. This allows Astro to better support single-framework usage, as well as advanced features like React Fast Refresh.

#### What should I do?

If you are using multiple JSX frameworks in the same project, set `include` (and optionally `exclude`) to an array of files and/or folders. Wildcards may be used to include multiple file paths.

We recommend placing common framework components in the same folder (e.g. `/components/react/` and `/components/solid/`) to make specifying your includes easier, but this is not required:

> ```js

Reviewer 6: As with other examples of the page, it would be better to highlight the added code for consistency.

```suggestion
```js ins={11-21}
```

Reviewer: Oooh, good call @morinokami! Since the integrations themselves will already be there, I think maybe JUST highlighting the `include` lines make sense? If you update, I'll commit!

Reviewer: And, OF COURSE, @morinokami you are absolutely allowed to review this! You are one of our most eagle-eyed contributors! It's not complete until you say so! :purple_heart:

Reviewer 6: Updated my suggestion!!

````suggestion
```js ins={13,16,19}
````

import { defineConfig } from 'astro/config';
import preact from '@astrojs/preact';
import react from '@astrojs/react';
import svelte from '@astrojs/svelte';
import vue from '@astrojs/vue';
import solid from '@astrojs/solid-js';

export default defineConfig({
  // Enable many frameworks to support all different kinds of components.
  // No `include` is needed if you are only using a single framework!
  integrations: [
		preact({
			include: ['**/preact/*']
		}),
		react({
			include: ['**/react/*']
		}),
		solid({
			include: ['**/solid/*'],
		}),
  ]
});
```

### Changed: `Astro.cookies.get(key)` can return `undefined`

In Astro v2.x, [`Astro.cookies.get(key)`](/en/reference/api-reference/#astrocookies) would always return an [`AstroCookie` object](/en/reference/api-reference/#astrocookie), even if the cookie did not exist. To check for its existence, you needed to use `Astro.cookies.has(key)`.

Astro v3.0 returns `undefined` for `Astro.cookies.get(key)` if the cookie does not exist.

#### What should I do?

> This change will not break existing code that checks for the existance of the `Astro.cookie` object before using `Astro.cookies.get(key)`, but is now no longer required.

Reviewer 3: I believe `existance` doesn't exist, but fixing the spelling got me a bit like "we have existing and existence in the same sentence??" - so perhaps we could do a bit more tweaking to avoid the similar words.
```suggestion
This change will not break any code that checks for the existence of the `Astro.cookie` object before using `Astro.cookies.get(key)`, but is now no longer required.
```

You can safely remove any code that uses `has()` to check if the value of `Astro.cookies` is `undefined`:

```js del={1-3} ins={5-7}
if (Astro.cookies.has(id)) {
  const id = Astro.cookies.get(id)!;
}

const id = Astro.cookies.get(id);
if (id) {
}
```

### Changed: internal Astro API entry point export paths

In Astro v2.x, you could import internal Astro APIs from `astro/internal/*` and `astro/runtime/server/*`.

> Astro v3.0 removes the two entry points in favour of the existing `astro/runtime/*` entrypoint. Additionally, a new `astro/compiler-runtime` export has been added for compiler-specific runtime code.

Reviewer 3: 🚓 British English Police 👮

also, there is entry point and entrypoint, I believe the separated one is for, well, saying "entry point" and "entrypoint" is kind of the technical thingy we use, pointing out in case these should actually be the same.
```suggestion
Astro v3.0 removes the two entry points in favor of the existing `astro/runtime/*` entrypoint. Additionally, a new `astro/compiler-runtime` export has been added for compiler-specific runtime code.
```

Reviewer: CANADIAN English Police, but yes! :salute:

#### What should I do?

These are entry points for Astro's internal API and should not affect your project. But if you do use these entrypoints, update as shown below:

>  ```js del={1,4,10} ins={2,5,11}
>  import 'astro/internal/index.js';
>  import 'astro/runtime/server/index.js';
> 
>  import 'astro/server/index.js';
>  import 'astro/runtime/server/index.js';
> ```
> 
>  ```js ins={5} del={4}
> import { transform } from '@astrojs/compiler';
> 
> const result = await transform(source, {
> internalURL: 'astro/runtime/server/index.js',
> internalURL: 'astro/compiler-runtime',
>   // ...
> });
>  ```

Reviewer 6: 🚓 Indentation Police 👮

````suggestion
```js del={1,4,10} ins={2,5,11}
import 'astro/internal/index.js';
import 'astro/runtime/server/index.js';

import 'astro/server/index.js';
import 'astro/runtime/server/index.js';
```

```js ins={5} del={4}
import { transform } from '@astrojs/compiler';

const result = await transform(source, {
  internalURL: 'astro/runtime/server/index.js',
  internalURL: 'astro/compiler-runtime',
  // ...
});
```
````

export default defineConfig({
  integrations: [
    image(),
  ]
})
```

### Removed: `<Markdown />` component

In Astro v1.x, Astro deprecated the `<Markdown />` component and moved it to an external package.

Astro v3.0 completely removes the package `@astrojs/markdown-component`. Astro's `<Markdown />` component will no longer work in your project.

#### What should I do?

Remove all instances of the `@astrojs/markdown-component`.

```astro del={2} title="src/components/MyAstroComponent.astro"
---
import Markdown from '@astrojs/markdown-component';
---
```

To continue using a similar `<Markdown />` component in your code, consider using [community integrations](https://astro.build/integrations/) such as [`astro-remote`](https://github.com/natemoo-re/astro-remote). Be sure to update your `<Markdown />` component imports and attributes as necessary, according to the integration's own documentation. 

Otherwise, delete all references to importing Astro's `<Markdown />` component and the component itself in your `.astro` files. You will need to rewrite your content as HTML directly or [import Markdown](/en/guides/markdown-content/#importing-markdown) from a `.md` file.

> ### Removed: `build.excludeMiddleware` and `build.split`

Reviewer 7: ```suggestion
### Deprecated: `build.excludeMiddleware` and `build.split`
```

That's on me, apologies. These options are still available.

Astro v3.0 no longer appends `import.meta.env.BASE_URL` with a trailing slash by default, nor when `trailingSlash: "ignore"` is set. (The existing behavior of `base` in combination with `trailingSlash: "always"` or `trailingSlash: "never"` is unchanged.)

#### What should I do?

If your `base` already has a trailing slash, no change is needed.

If your `base` does not have a trailing slash, add one if you wish to preserve the previous default (or `trailingSlash: "ignore"`) behavior:

```js title="astro.config.mjs" del={4} ins={5}
import { defineConfig } from "astro/config";

export default defineConfig({
  base: 'my-base',
  base: 'my-base/',
});
```

### Changed default: `compressHTML`

In Astro v2.x, Astro only compressed your emitted HTML when [`compressHTML`](/en/reference/configuration-reference/#compresshtml) was explicitly set to `true`. The default value was `false`.

Astro v3.0 now compresses emitted HTML by default.

#### What should I do?

You can now remove `compressHTML: true` from your configuration as this is the new default behavior.

```js title="astro.config.mjs" del={4}
import { defineConfig } from "astro/config";

export default defineConfig({

> 	compressHTML: true

Reviewer 7: ```suggestion
  compressHTML: true
```

})
```

You must now set `compressHTML: false` to opt out of HTML compression.

### Changed default: `scopedStyleStrategy`

In Astro v2.x, the default value of [`scopedStyleStrategy`](/en/reference/configuration-reference/#scopedstylestrategy) was `"where"`.

Astro v3.0 introduces a new, default value: `"attribute"`. By default, styles are now applied using `data-*` attributes.

#### What should I do?

To retain your project's current [style scoping](/en/guides/styling/#scoped-styles), update the configuration file to the previous default value: 

```js title="astro.config.mjs" ins={4}
import { defineConfig } from "astro/config";

export default defineConfig({

> 	scopedStyleStrategy: "where"

Reviewer 7: ```suggestion
  scopedStyleStrategy: "where"
```

})
```

### Changed default: `inlineStyleSheets`

> In Astro v2.x, all project stylesheets were sent as link tags by default. You could opt in to inlining them into `<style>` tags everytime with `"always"`, or to inlining only stylesheets below a certain size with `"auto"` by setting the [`build.inlineStylesheets`](/en/reference/configuration-reference/#buildinlinestylesheets) configuration. The default setting was `"never"`.

Reviewer 8: Please write "every time" as two words. As one word, it would become an adjective and we don't use it nearly as often as we use the adjective "everyday".

> In Astro v2.x, all project stylesheets were sent as link tags by default. You could opt in to inlining them into `<style>` tags everytime with `"always"`, or to inlining only stylesheets below a certain size with `"auto"` by setting the [`build.inlineStylesheets`](/en/reference/configuration-reference/#buildinlinestylesheets) configuration. The default setting was `"never"`.

Reviewer 8: ```suggestion
In Astro v2.x, all project stylesheets were sent as link tags by default. You could opt in to inlining them into `<style>` tags every time with `"always"`, or to inlining only stylesheets below a certain size with `"auto"` by setting the [`build.inlineStylesheets`](/en/reference/configuration-reference/#buildinlinestylesheets) configuration. The default setting was `"never"`.
```

Please write "every time" as two words. As one word, it would become an adjective and we don't use it nearly as often as we use the adjective "everyday" (as in "Using 'everytime' as an adjective is not an everyday occurrence.").

Astro v3.0 now requires you to specify which framework to use for your files with new `include` and `exclude` integration config options when you have multiple JSX framework integrations installed. This allows Astro to better support single-framework usage, as well as advanced features like React Fast Refresh.

#### What should I do?

If you are using multiple JSX frameworks in the same project, set `include` (and optionally `exclude`) to an array of files and/or folders. Wildcards may be used to include multiple file paths.

We recommend placing common framework components in the same folder (e.g. `/components/react/` and `/components/solid/`) to make specifying your includes easier, but this is not required:

```js
import { defineConfig } from 'astro/config';
import preact from '@astrojs/preact';
import react from '@astrojs/react';
import svelte from '@astrojs/svelte';
import vue from '@astrojs/vue';
import solid from '@astrojs/solid-js';

export default defineConfig({
  // Enable many frameworks to support all different kinds of components.
  // No `include` is needed if you are only using a single framework!
  integrations: [

> 		preact({
> 			include: ['**/preact/*']
> 		}),
> 		react({
> 			include: ['**/react/*']
> 		}),
> 		solid({
> 			include: ['**/solid/*'],
> 		}),

Reviewer 7: ```suggestion
    preact({
      include: ['**/preact/*']
    }),
    react({
      include: ['**/react/*']
    }),
    solid({
      include: ['**/solid/*'],
    }),
```

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import FileTree from '~/components/FileTree.astro'

This guide will help you migrate from Astro v2 to Astro v3.

Need to upgrade an older project to v2? See our [older migration guide](/en/guides/upgrade-to/v2/).

## Upgrade Astro

Update your project's version of Astro to the latest version using your package manager. If you're using Astro integrations, please also update those to the latest version.

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  # Upgrade to Astro v3.x
  npm install astro@latest
  
  # Example: upgrade React and Tailwind integrations
  npm install @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  # Upgrade to Astro v3.x
  pnpm install astro@latest

  # Example: upgrade React and Tailwind integrations
  pnpm install @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  # Upgrade to Astro v3.x
  yarn add astro@latest
  
  # Example: upgrade React and Tailwind integrations
  yarn add @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
</PackageManagerTabs>

> ## Astro v3.0 Experimental Flags Removed

Reviewer: ```suggestion
:::note[Need to continue?]
After upgrading Astro to the latest version, you may not need to make any changes to your project at all!

But, if you notice errors or unexpected behavior, please check below for what has changed that might need updating in your project.
:::

## Astro v3.0 Experimental Flags Removed
```

Otherwise, delete all references to importing Astro's `<Markdown />` component and the component itself in your `.astro` files. You will need to rewrite your content as HTML directly or [import Markdown](/en/guides/markdown-content/#importing-markdown) from a `.md` file.

### Removed: deprecated 1.x APIs

In Astro v1.x, Astro deprecated config options as well as script/style `global` and `hoist` attributes types. These were kept for backwards compatibility.

Astro v3.0 removes these deprecated APIs entirely. They can no longer be used in your Astro project.

#### What should I do?

If you are continuing to use v1.x APIs, use the new APIs for each feature instead:

- Deprecated config options: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-configuration-api)
- Deprecated script/style attribute types: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-default-script-behavior)

### Removed: `image` from `astro:content` in content collections schema

In Astro 2.x, the content collections API deprecated an `image` export from `astro:content` for use in your content collections schemas.

Astro 3.0 removes this export entirely.

#### What should I do?

> If you are using the deprecated `image()` from `astro:content`, remove it as this no longer exists. Validate images through [the `image` helper from `schema`](/en/guides/images/#update-content-collections-schemas) instead:

Reviewer 3: Point out that this link isn't marked as broken, but this specific heading doesn't exist yet! till the v3 images page is merged

Reviewer: Thanks, Yan! There will be some like this, since some headings on the new page are in fact the same as the ones on the existing page!

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  # Upgrade to Astro v3.x
  npm install astro@latest
  
  # Example: upgrade React and Tailwind integrations
  npm install @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  # Upgrade to Astro v3.x
  pnpm install astro@latest

  # Example: upgrade React and Tailwind integrations
  pnpm install @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  # Upgrade to Astro v3.x
  yarn add astro@latest
  
  # Example: upgrade React and Tailwind integrations
  yarn add @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
</PackageManagerTabs>

## Astro v3.0 Experimental Flags Removed

Remove the following experimental flags from `astro.config.mjs`:

```js del={5-8}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
    assets: true,
    viewTransitions: true,
  },
})
```

These features are now available by default:

- [View Transitions](/en/guides/view-transitions/) for animated page transitions and persistent islands.

> - A new image services API `astro:assets` for [using images in Astro](/en/guides/images/), including a new `<Image />` component and `getImage()` function.

Reviewer: ```suggestion
- A new image services API `astro:assets` for [using images in Astro](/en/guides/images/), including a new `<Image />` component and `getImage()` function. Please read the detailed [image upgrade advice](/en/guides/images/#upgrade-to-v30-from-v2x) to see how this might affect your project.
```

  integrations: [
    image(),
  ]
})
```

### Removed: `<Markdown />` component

In Astro v1.x, Astro deprecated the `<Markdown />` component and moved it to an external package.

Astro v3.0 completely removes the package `@astrojs/markdown-component`. Astro's `<Markdown />` component will no longer work in your project.

#### What should I do?

Remove all instances of the `@astrojs/markdown-component`.

```astro del={2} title="src/components/MyAstroComponent.astro"
---
import Markdown from '@astrojs/markdown-component';
---
```

To continue using a similar `<Markdown />` component in your code, consider using [community integrations](https://astro.build/integrations/) such as [`astro-remote`](https://github.com/natemoo-re/astro-remote). Be sure to update your `<Markdown />` component imports and attributes as necessary, according to the integration's own documentation. 

Otherwise, delete all references to importing Astro's `<Markdown />` component and the component itself in your `.astro` files. You will need to rewrite your content as HTML directly or [import Markdown](/en/guides/markdown-content/#importing-markdown) from a `.md` file.

### Removed: deprecated 1.x APIs

> In Astro v2.x, deprecated config options as well as script/style `global` and `hoist` attributes types. These were kept for backwards compatibility.

Reviewer 2: ```suggestion
In Astro v1.x, Astro deprecated config options as well as script/style `global` and `hoist` attributes types. These were kept for backwards compatibility.
```

I think `Astro v1.x` matches the heading here.

Astro v3.0 removes these deprecated APIs entirely. They can no longer be used in your Astro project.

#### What should I do?

If you are continuing to use v1.x APIs, use the new APIs for each feature instead:

- Deprecated config options: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-configuration-api)
- Deprecated script/style attribute types: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-default-script-behavior)

### Removed: `image` from `astro:content` in content collections schema

In Astro 2.x, the content collections API deprecated an `image` export from `astro:content` for use in your content collections schemas.

Astro 3.0 removes this export entirely.

#### What should I do?

If you are using the deprecated `image()` from `astro:content`, remove it as this no longer exists. Validate images through [the `image` helper from `schema`](/en/guides/images/#update-content-collections-schemas) instead:

 ```ts title="astro.config.mjs" del={1} ins={2} "({ image })"

> import { defineCollection, z, image() } from "astro:content";

Reviewer 2: ```suggestion
import { defineCollection, z, image } from "astro:content";
```

Otherwise, delete all references to importing Astro's `<Markdown />` component and the component itself in your `.astro` files. You will need to rewrite your content as HTML directly or [import Markdown](/en/guides/markdown-content/#importing-markdown) from a `.md` file.

### Removed: deprecated 1.x APIs

In Astro v1.x, Astro deprecated config options as well as script/style `global` and `hoist` attributes types. These were kept for backwards compatibility.

Astro v3.0 removes these deprecated APIs entirely. They can no longer be used in your Astro project.

#### What should I do?

If you are continuing to use v1.x APIs, use the new APIs for each feature instead:

- Deprecated config options: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-configuration-api)
- Deprecated script/style attribute types: See [the 0.26 migration guide](/en/guides/upgrade-to/v1/#new-default-script-behavior)

### Removed: `image` from `astro:content` in content collections schema

> In Astro 2.x, the content collections API deprecated an `image` export from `astro:content` for use in your content collections schemas.

Reviewer 9: ```suggestion
In Astro v2.x, the content collections API deprecated an `image` export from `astro:content` for use in your content collections schemas.
```

> Astro 3.0 removes this export entirely.

Reviewer 9: ```suggestion
Astro v3.0 removes this export entirely.
```

```astro title="src/components/MyAstroComponent.astro" del={4} ins={5}
<Component class:list={[
  'a',
  'b',
  new Set(['c', 'd'])
  ['c', 'd'] 
]} />
```

### Removed: passing `class:list` as a prop

In Astro v2.x, [`class:list` values](/en/reference/directives-reference/#classlist) were sent to components via [`Astro.props['class:list']`](/en/reference/api-reference/#astroprops).

Astro v3.0 normalizes `class:list` values into a string before being sent to components via `Astro.props['class']`

#### What should I do?

Remove any code that expects to receive the `class:list` prop.

```astro title="src/components/MyAstroComponent.astro" del={2,3,7} ins={4,8} "classList" "'class:list': classList"
---
import { clsx } from 'clsx';
const { class: className, 'class:list': classList } = Astro.props;
const { class: className } = Astro.props;
---
<div
  class:list={[className, classList]}
  class:list={[className]}
/>
```

### Removed: kebab-case transform for camelCase CSS variables

In Astro v2.x, camelCase [CSS variables](/en/guides/styling/#css-variables) passed to the `style` attribute were rendered as both camelCase (as written) and kebab-case (kept for backwards compatibility).

> Astro 3.0 removes the kebab-case transform for these camelCase CSS variable names, and only the original camelCase CSS variable is rendered.

Reviewer 9: ```suggestion
Astro v3.0 removes the kebab-case transform for these camelCase CSS variable names, and only the original camelCase CSS variable is rendered.
```

In Astro v2.x, the `markdown.drafts` configuration allowed you to have draft pages that were available in when running the dev server, but not built in production.

Astro v3.0 deprecates this feature in favor of the content collections method of handling draft pages by filtering manually instead, which gives more control over the feature.

#### What should I do?

To continue to mark some pages in your project as drafts, [migrate to content collections](/en/guides/content-collections/#migrating-from-file-based-routing) and [manually filter out pages](/en/guides/content-collections/#filtering-collection-queries) with the `draft: true` frontmatter property instead.

### Deprecated: returning simple object in endpoints

In Astro v2.x, endpoints could return a simple object, which would be converted to a JSON response.

Astro v3.0 deprecates this behavior in favor of returning a `Response` object directly.

#### What should I do?

Update your endpoints to return a `Response` object directly.

```ts title="endpoint.json.ts" del={2} ins={3}
export async function GET() {
  return { body: { "title": "Bob's blog" }};
  return new Response(JSON.stringify({ "title": "Bob's blog" }));
}
```

> If you really need to keep previous format, you can use the `ResponseWithEncoding` object, but will be depricated in the future.

Reviewer 3: Wording nit
```suggestion
If you really need to keep the previous format, you can use the `ResponseWithEncoding` object but will be deprecated in the future.
```
